File: java/Exercises.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.IntStream;

public class Exercises{

    public static List<Integer> change(int amount) {
        if(amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative.");
        }
        var coins = new ArrayList<Integer>();
        for (var coin : List.of(25, 10, 5, 1)) {
            coins.add(amount / coin);
            amount %= coin;
        }
        return List.copyOf(coins);
    }

    public static String stretched(String s) {
        String result = "";
        var characters = s.replaceAll("\\s+", "").codePoints().toArray();
        for(var i = 0; i < characters.length; i++){
            result += String.copyValueOf(Character.toChars(characters[i])).repeat(i + 1);
        }
        return result;
    }

    public static IntStream powers(int base) {
        return IntStream.iterate(1, n -> n * base);
    }

    // Word record for say
    record Word(String word){
        String ok() {
            return word;
        }

        Word and(String nextWord) {
            return new Word(word + " " + nextWord);
        }       
    }

    // say() method if no argument is passed through.
    public static String say() {
        return "";

    }

    // say() method if an argument is passed through.
    public static Word say(String word) {
        return new Word(word);
    }

    public static Optional<String> findFirstThenLower(Predicate<String> p, List<String> strings){
        return strings.stream().filter(p).findFirst().map(string -> string.toLowerCase());    
    }

    public static List<String> topTenScorers(Map<String, List<String>> stats) {
        record Player(String name, double ppg, String team) {}
        var playerList = new ArrayList<Player>();

        stats.keySet()
        .forEach(team -> stats.get(team)
        .forEach(profile -> { 
            var playerStats = profile.split(",");
            var playerName = playerStats[0];
            var gamesPlayed = Double.parseDouble(playerStats[1]);
            var ppg = Double.parseDouble(playerStats[2]) / gamesPlayed;
            if(gamesPlayed >= 15){
                playerList.add(new Player(playerName, ppg, team));
            }
        ;}));

        return playerList.stream()
            .sorted((playerOne, playerTwo) -> Double.compare(playerOne.ppg(), playerTwo.ppg()))
            .map(player -> String.format("%s|%,.2f|%s", player.name, player.ppg, player.team))
            .limit(10)
            .toList();
    }
}

File: python/exercises.py
import math

# Question 1
def change(cents):

    # error and type checking
    if (cents - int(cents) != 0):
        error = TypeError('No fractional amounts allowed')
        raise error
    
    if cents < 0:
        error = ValueError('amount cannot be negative')
        raise error
    
    # calculating the number of coins needed
    quarters = math.floor(cents / 25)
    cents %= 25

    dimes = math.floor(cents / 10)
    cents %= 10

    nickels = math.floor(cents / 5)
    cents %= 5

    pennies = math.floor(cents / 1)
    cents %= 1

    
    return (quarters, dimes, nickels, pennies)


# Question 2
def stretched(s):

    # removing white space and possible hiccups to the code
    stripped = s.replace(" ", "")
    stripped = stripped.replace('\n', '')
    stripped = stripped.replace('\t', '')
  
    # turing the code into a list for easier manipulation
    characters = list(stripped)

    # iterating through list and strecthing it 
    for i in range (len(characters)):
       
        characters[i] = characters[i] * (i+1)

    
    # turning it back into a string
    s = "".join(characters)
    

    return s


# Question 3
def powers(*,base, limit):
    power = 1

    # loop to make sure we do not go over limit
    while(power <= limit):
        yield power
        power = power * base

      
# Question 4
def say(s=None):
    if s is None:
        return ''
    
    # defining another function to keep track of previous words
    def inner_say(in_s = None):
        if in_s is None:
            return s
        return say('{} {}'.format(s, in_s))

    return inner_say

# Question 5
def find_first_then_lower(predicate, strings):
    for word in strings:
        if predicate(word):
            return word.lower()
    return None

# Question 6
def top_ten_scorers(stats):
    validPlayers = []
    topPlayers = []
    for team in stats:
        for players in stats[team]:
            if players[1] >= 15:
                name = players[0]
                ppg = format(players[2] / players[1], ".2f")
                validPlayers.append((name, ppg, team))
    validPlayers.sort(key = lambda ppg : -float(ppg[1]))
    for player in validPlayers:
        playerData = ""
        playerData = playerData + "|".join(player)
        topPlayers.append(playerData)

    return topPlayers[:10]

# Question 7
def turing_machine(machine, input):
    return []


